### ISSUE TITLE
[DSLX/IR/Codegen] Inlining IR with inter proc channels fails, fifo_depth must be < 1 but was  (none)

### ISSUE THREAD

**Issue by mtdudek** [google/xls#976]
I've encountered an issue with a design that uses 2 procs.
The first proc receives data from external channel, does some processing and
it sends data to the second proc. Second proc also does some processing and
it sends its results to an external channel.

During code generation I get error:
```
Only streaming channels of FIFO depth one or less are supported in proc inlining. Channel `proc1_proc2` has depth: (none)
```
 I've modified DLSX to IR converter, so that every intra DLSX channel has depth of 1.
It was enough to get codegen working but as i can see in #950 multiproc codegen is being developed.
I think this error is similar to #708.

I'll create PR with my current fix.
/cc @proppy

PR: https://github.com/google/xls/pull/977

**Comment by grebe** [comment]
I don't think we want fifo_depth to be hardcoded to any value, and none is the correct default value.

Instead, I think it would be better for internal channels to optionally set the fifo depth. For example:

```
let (chan0_p, chan0_c) = chan<F32>();  // fifo depth 'none' by default
let (chan1_p, chan1_c) = chan<F32>(fifo_depth=3);
```

**Comment by mtdudek** [comment]
I don't know, if adding fifo_depth to DSLX is a great idea.
From DSLX point of view channels are non-blocking(#965),
so they can be seen as infinite. 
How about adding flag to ir_converter that can set inter proc fifo_depth?
Also if `none` is a valid value for fifo_depth, how it should be handled by proc inline?

**Comment by grebe** [comment]
I think `none` should be interpreted as 'unspecified' (essentially infinite) and the compiler should be free to insert FIFOs of whatever size it likes as long as it doesn't cause deadlock. In practice, we don't currently insert those FIFOs. Proc inlining won't inline channels w/ fifo depth of `none`.

I think a flag that sets all internal channels' fifo depths is not ideal- internal channel names are very brittle, you probably don't want them to all have the same depth, and I think in general we have too many flags and not enough source-level configuration.

It seems reasonable to me to set fifo depth in the source (potentially via parameter), and if you don't want to have to specify it everywhere have a pass that does analysis (or maybe just blanket applies a configurable default) to any unspecified depths.

Note also that proc inlining won't inline anything w/ depth > 1. I've started some exploration on a pass that will lower depths > 1 into a FIFO proc and set the channel to have depth 0 allowing inlining to work for depth > 1 (there are some issues, notably proc inlining can't inline nonblocking receive).

**Comment by grebe** [comment]
Yeah, agreed! I was playing around with different approaches and I landed on essentially the same idea. The referenced commit adds that functionality.

You _should_ be able to do parametric expressions, e.g. `let (out, in) = chan<F32, {A+B+C}>;`, although for now I imagine we want to stick to 0 or 1 for the most part so that's not too useful.

**Comment by mtdudek** [comment]
I've also done some work with parser here #987, but I see I was to late :)

**Comment by grebe** [comment]
Ah, I missed that! Sorry for the overlap

**Comment by mtdudek** [comment]
I don't which one would we like to keep, but as yours is already on main. I think my should be rejected?

**Comment by grebe** [comment]
I think so?

I've been reviewing the PR and it looks like our approaches were slightly different- I added fifo depth to the channel decl construct (`let (p, c) = chan<...>;`) whereas you added it to the channel type. My thinking was that we should only set fifo depths on internal channels, which are made via channel decls, and that it is less of a property of the type and more a property of the decl. Having gone down the path you did, do you think I'm missing anything- are there things you can think of that you'd want to do with fifo depth on the type rather than the decl?

**Comment by mtdudek** [comment]
I haven't though about how can we use depth in type definition, but I think there is use case than can benefit.
Let say user knows that they will required fifo depth of `x`, as they will be sending and/or receiving in bursts.
If they can put it in type like: `in_: chan<F32, x>;`. All FIFOs of depth < x would result in type checking error.

You are right, there aren't many use cases for having depth in type declaration.

**Comment by proppy** [comment]
@mtdudek my understanding was that this is blocking #994 and #995?

@grebe's commit suggested more work is needed, so it makes sense to leave this bug open, but I'm curious if the addition of fifo depth in 9939db01b6a59ac8c226e25a313b9212b859ddd8 is enough to unblock you
